src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs

```diff
@@ -41,17 +42,27 @@ public object GetService(Type serviceType)
 
         internal object CaptureDisposable(object service)
         {
-            Debug.Assert(!_disposed);
-
             _captureDisposableCallback?.Invoke(service);
 
             if (ReferenceEquals(this, service) || !(service is IDisposable || service is IAsyncDisposable))
             {
                 return service;
             }
 
-            lock (ResolvedServices)
+            lock (_lock)
             {
+                if (_disposed)
+                {
+                    if (service is IAsyncDisposable asyncDisposable)
+                    {
+                        // TODO complete
```

Comment:
- [ ] TODO complete

Reply:
Oh this is tricky.... since this code path is all synchronous.

Reply:
We may have to block here... We should also maybe prefer IDisposable over IAsyncDisposable to make this less of a problem.